### ISSUE TITLE
[COLOR] Olive Fails To Handle Negative Colours Properly

### ISSUE THREAD

**Issue by ThomasWilshaw** [olive-editor/olive#1876]
**Commit Hash** 
1e3e620

**Platform** 
Windows 10

**Summary**
If a negative colour is run through the view transform it can sometimes cause strange colours to appear.

**Additional Information / Output**
If you load the attached image (it's zipped as Github won't support EXR) into Olive and set it's Input Color Space to be Alexa Wide Gamut EI 800 the image should look blue ish. If you then set the Display View to Filmic and Base Constrast the image skews magenta which is not correct. Using the same config in Blender gives the expected result of a slighty different blue implying this is an Olive issue.

After the Alexa input transform the image in reference space has a negative red value
![image](https://user-images.githubusercontent.com/7781683/157496968-92e112c6-f64a-4ca9-9be5-355a2b245c63.png)

If I add a small amount of red to the reference space via an Math node the image returns to the blueish colour we would expect. I also found that if I clip negative values in the FIlmic Log Encoding space via the config the issue goes away. These both imply that the issue is with how Olive handles negative values.
[negative.zip](https://github.com/olive-editor/olive/files/8216845/negative.zip)

![image](https://user-images.githubusercontent.com/7781683/157498581-f3d73938-ee77-4809-9b9d-9770d39c043d.png)

**Comment by itsmattkc** [comment]
It looks OpenGL specific since the color in the pixel sampler (which is CPU generated) appears correct. What happens if you clamp the values in the viewer?

**Comment by ThomasWilshaw** [comment]
I'm not sure if this is exacty what you meant but I've added this line to `colomanage.frag`:
`col.rgb = max(col.rgb, vec3(0.0));`
before we do any processing and it seems to have solved the problem. Does this point towards it being a problem with how OCIO handles negatives in its GLSL code maybe?

**Comment by sobotka** [comment]
> Does this point towards it being a problem with how OCIO handles negatives in its GLSL code maybe?

OCIO strives for invertibility, always.

The part that is confusing to me here is how a clip at the display somehow makes thing suddenly work. If the negative values are shipped out to display, the display implicitly clamps them.

That is, it is probably worth considering where the negative values are manifesting the magenta, which is probably obvious to someone with a larger chunk of grey matter than I have.

**Comment by itsmattkc** [comment]
> I'm not sure if this is exacty what you meant but I've added this line to `colomanage.frag`: `col.rgb = max(col.rgb, vec3(0.0));` before we do any processing and it seems to have solved the problem. Does this point towards it being a problem with how OCIO handles negatives in its GLSL code maybe?

Try doing the clamp after the processing, preferably just before it goes out to display. I think it might simply be "OpenGL doesn't know how to display a negative RGB value" rather than anything necessarily OCIO related.

**Comment by ThomasWilshaw** [comment]
That did also work yes. The thing is though this is the first thing the OCIO function does:
```
  {
    vec3 minValue = vec3(1.17549435e-38, 1.17549435e-38, 1.17549435e-38);
    vec3 lin_slope = vec3(1., 1., 1.);
    vec3 lin_offset = vec3(0.0039215683937072754, 0.0039215683937072754, 0.0039215683937072754);
    vec3 log_slope = vec3(1.44269502, 1.44269502, 1.44269502);
    vec3 log_offset = vec3(0., 0., 0.);
    outColor.rgb = max( minValue, (outColor.rgb * lin_slope + lin_offset) );
    outColor.rgb = log_slope * log(outColor.rgb ) + log_offset;
  }
```
So there should never be a negative there anyway surely?

**Comment by sobotka** [comment]
Out of curiosity @ThomasWilshaw does the recent bump to OpenGL ES 3.2 as per @itsmattkc change any handling?

**Comment by ThomasWilshaw** [comment]
No I'm still having the same issue on the ES 3.2 branch unfortunately

**Comment by ThomasWilshaw** [comment]
Matt has solved it (43cef6e332039059f8e82454b82184a6be92a221). 3D textures were wrapping around so a negative value was pulling the wrong value from the texture, This has fixed the issue and explains why we were getting those colours but I'm still not sure how a negative was getting through.
